facebookresearch/ParlAI
# Add Taskmaster-1 dataset

**Patch description**
Adds Google AI's [TaskMaster-1](https://ai.google/tools/datasets/taskmaster-1) dataset.

**Testing steps**
Enter steps to test your pull request. Give a clear and concise description of
what you expected to happen during testing.
Run the following commands to view how many episodes and examples were loaded from the data
```
python examples/display_data.py -t taskmaster:SelfDialogueTeacher
python examples/display_data.py -t taskmaster:WozDialogueTeacher
python examples/display_data.py -t taskmaster:SelfDialogueSegmentTeacher
```
The above commands also work if you append any of: ```-dt train```, ```-dt valid```, ```-dt test``` to them. Since there is no formal split defined as of now for the data, the outputs remain unchanged. If you would like to see examples in the first two commands above in which the roles are reversed (i.e. label is the USER text) simply increase the number of examples to be shown (eg. ``` -n 1000000```
The expected output can be seen in logs.

**Logs**
- ***SelfDialogueTeacher***

USER text as input ASSISTANT text as label
```
- - - - - - - - - - - - - - - - - - - - -
~~
[taskmaster:SelfDialogueTeacher]: Hi, I would like to book a table for a restaurant.
[labels: Where would you like to book?]
~~
[taskmaster:SelfDialogueTeacher]: I would like to book for Sofia in Englewood, NJ.
[labels: For how many people?]
~~
```
ASSISTANT text as input USER text as label
```
- - - - - - - - - - - - - - - - - - - - -
~~
[taskmaster:SelfDialogueTeacher]: Absolutely, I can help you with that. How many people are in your party? And what date are you looking at?
[labels: April 20th for 6 people please.]
~~
[taskmaster:SelfDialogueTeacher]: Of course, let me look that up for you. What time would you like?
[labels: 7pm]
~~
[taskmaster:SelfDialogueTeacher]: Excellent. And do you have any special requirements for this reservation?
[labels: Yes, I need at table that can accommodate a wheelchair.]
~~
[taskmaster:SelfDialogueTeacher]: Okay, let me have a look here and see what we can find for you.
[labels: Okay thank you.]
~~
[taskmaster:SelfDialogueTeacher]: Absolutely. We have that date and time available, but the table cannot accommodate a wheelchair. Is there a different date or restaurant you want to try?
[labels: Okay, how about Italian Garden?]
~~
[taskmaster:SelfDialogueTeacher]: All right, let me see what we have
[labels: Okay]
~~
[taskmaster:SelfDialogueTeacher]: So it looks like we have the table, but not at 7pm. We can try a different time, maybe 6pm or 8pm?
[labels: No, I am looking for something at 7pm.]
~~
```
Data info from output: ``` [ loaded 15416 episodes with a total of 161761 examples ]```

- ***WozDialogueTeacher***

USER text as input ASSISTANT text as label
```
- - - - - - - - - - - - - - - - - - - - -
~~
[taskmaster:WozDialogueTeacher]: Hi. I'd like to order some pizza from Bella Luna.
[labels: ok, how many pizzas would you like?]
~~
[taskmaster:WozDialogueTeacher]: I'm feeling like two medium pizzas with extra cheese on both of them.
[labels: sure. do you want two cheese pizzas or would you like to add different toppings?]
~~
[taskmaster:WozDialogueTeacher]: No, I Let's Let's start over. I'd like two medium pizzas. One with Canadian bacon. And the other, I would like anchovies.
Thin crust will be fine. And my my daughter likes for you to add extra cheese on both the pizza.
[labels: sure, so your order is: 2 medium thin pizzas, one with canadian bacon and the other with anchovies. extra cheese on both, is this correct?]
~~
[taskmaster:WozDialogueTeacher]: Yeah, that's perfect.
[labels: ok, the total is $38.]
~~
[taskmaster:WozDialogueTeacher]: Well, okay. Yeah, I guess that's in my budget.
[labels: your pizzas will be ready for pickup at bela luna's in 25 minutes.]
~~
[taskmaster:WozDialogueTeacher]: Okay.
[labels: enjoy, bye!]
```
ASSISTANT text as input USER text as label
```
- - - - - - - - - - - - - - - - - - - - -
~~
[taskmaster:WozDialogueTeacher]: hi, how can i assist you?
hello?
[labels: Hi. Yes, I would like a Uber.]
~~
[taskmaster:WozDialogueTeacher]: sure, what is the pickup location?
[labels: And the pickup will be
Lux in L.A.]
~~
[taskmaster:WozDialogueTeacher]: where would you like to be dropped off?
[labels: (deleted)
(deleted)
I would like to be dropped off at the Hyatt Regency Los Angeles airport hotel.]
~~
[taskmaster:WozDialogueTeacher]: what kind of ride would you like?
[labels: Just the Uber X.]
~~
[taskmaster:WozDialogueTeacher]: ok, how many people are riding with you?
you can only take up to 4 people on uberX. i suggest that you request uberXL for larger groups.
[labels: All right. Yeah, let me get a Uber XL van.]
~~
[taskmaster:WozDialogueTeacher]: sure. your fare estimate is $12.69, can i confirm this request?
[labels: Can I shedule for tonight at 9:00 p.m.?]
~~
[taskmaster:WozDialogueTeacher]: ok. your uberxl is scheduled for tonight at 9 pm.
[labels: Okay and
Thank you. I really do appreciate it.]
```
Data info from output: ```[ loaded 10714 episodes with a total of 102677 examples ]```

- ***SelfDialogueSegmentTeacher***
```
- - - - - - - - - - - - - - - - - - - - -
~~
[label_types]:
  ['movie_ticket.name.movie', 'movie_ticket.name.movie']
[taskmaster:SelfDialogueSegmentTeacher]: Hi I would like to see if the Movie What Men Want is playing here.
[labels: What Men Want]
~~
[label_types]:
  ['movie_ticket.num.tickets']
  ['movie_ticket.num.tickets']
[taskmaster:SelfDialogueSegmentTeacher]: Yes, for me and a friend so two tickets please
[labels: two|two tickets]
~~
[label_types]:
  ['movie_ticket.time.start', 'movie_ticket.time.start']
  ['movie_ticket.time.start', 'movie_ticket.time.start']
  ['movie_ticket.time.start', 'movie_ticket.time.start']
[taskmaster:SelfDialogueSegmentTeacher]: That movie is showing at 4, 5, and 8pm.
[labels: 4|5|8pm]
~~
[label_types]:
  ['movie_ticket.time.start']
[taskmaster:SelfDialogueSegmentTeacher]: Okay. Is there anymore movies showing around 8pm
[labels: 8pm]
~~
```
Data info from output: ```[ loaded 7691 episodes with a total of 68228 examples ]```


**Other information**
Teachers added:
- ```SelfDialogueTeacher```: Uses self-dialogs.json from the dataset. The task here is to predict the ***reply*** of the assistant to the user.
- ```WozDialogueTeacher```: Uses woz-dialogs.json from the dataset. The task here is to predict the ***reply*** of the assistant to the user. (Needs much more processing of data than the above file)
- ```SelfDialogueSegmentTeacher```: The task is to now predict the different segments from the text in self-dialogs.json.
Notes:
- Each teacher has a different meaning of what constitutes as an example and what as an episode and hence each of them calculates those numbers independently

**Data tests (if applicable)**
As requested by CircleCI I ran tests/datatests/test_new_tasks.py. The output is shown below:
```
$ python tests/datatests/test_new_tasks.py
.
----------------------------------------------------------------------
Ran 1 test in 520.613s

OK
```

Since I added a new teacher, as requested by PR template, I ran tests.suites.datatests. The output is shown below:
```
$ python setup.py test -s tests.suites.datatests
running test
running egg_info
writing parlai.egg-info/PKG-INFO
writing dependency_links to parlai.egg-info/dependency_links.txt
writing requirements to parlai.egg-info/requires.txt
writing top-level names to parlai.egg-info/top_level.txt
reading manifest file 'parlai.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'parlai.egg-info/SOURCES.txt'
running build_ext
test_verify_data (test_new_tasks.TestNewTasks) ... ok

----------------------------------------------------------------------
Ran 1 test in 518.258s

OK
```
